rust/crd/src/resource.rs

```diff
@@ -31,132 +78,262 @@ impl RoleResource {
+                max: Some(Quantity("4".to_owned())),
+            },
+            memory: MemoryLimitsFragment {
+                limit: Some(Quantity("2Gi".to_owned())),
+                runtime_limits: NoRuntimeLimitsFragment {},
+            },
+            storage: storage::DruidStorageFragment {},
+        };
+    pub static ref HISTORICAL_RESOURCES: ResourcesFragment<storage::HistoricalStorage, NoRuntimeLimits> =
+        ResourcesFragment {
+            cpu: CpuLimitsFragment {
+                min: Some(Quantity("200m".to_owned())),
+                max: Some(Quantity("4".to_owned())),
+            },
+            memory: MemoryLimitsFragment {
+                limit: Some(Quantity("2Gi".to_owned())),
+                runtime_limits: NoRuntimeLimitsFragment {},
+            },
+            storage: storage::HistoricalStorageFragment {},
+        };
+}
+
+fn default_resources(role: &DruidRole) -> Option<RoleResourceFragment> {
+    match role {
+        DruidRole::Historical => Some(RoleResourceFragment::HistoricalFragment(
+            HISTORICAL_RESOURCES.clone(),
+        )),
+        _ => Some(RoleResourceFragment::DruidFragment(
+            DEFAULT_RESOURCES.clone(),
+        )),
+    }
+}
+
+fn role_resources(druid: &DruidCluster, role: &DruidRole) -> Option<RoleResourceFragment> {
+    match role {
+        DruidRole::Broker => druid
+            .spec
+            .brokers
+            .clone()
+            .config
+            .config
+            .resources
+            .map(RoleResourceFragment::DruidFragment),
+        DruidRole::Coordinator => druid
+            .spec
+            .coordinators
+            .clone()
+            .config
+            .config
+            .resources
+            .map(RoleResourceFragment::DruidFragment),
+        DruidRole::Historical => druid
+            .spec
+            .historicals
+            .clone()
+            .config
+            .config
+            .resources
+            .map(RoleResourceFragment::HistoricalFragment),
+        DruidRole::MiddleManager => druid
+            .spec
+            .middle_managers
+            .clone()
+            .config
+            .config
+            .resources
+            .map(RoleResourceFragment::DruidFragment),
+        DruidRole::Router => druid
+            .spec
+            .routers
+            .clone()
+            .config
+            .config
+            .resources
+            .map(RoleResourceFragment::DruidFragment),
+    }
+}
+
+fn rolegroup_resources(
+    druid: &DruidCluster,
+    role: &DruidRole,
+    rolegroup_ref: &RoleGroupRef<DruidCluster>,
+) -> Option<RoleResourceFragment> {
+    match role {
+        DruidRole::Broker => druid
+            .spec
+            .brokers
+            .clone()
+            .role_groups
+            .get(&rolegroup_ref.role_group)
+            .map(|rg| &rg.config.config)
+            .and_then(|rg| rg.resources.clone())
+            .map(RoleResourceFragment::DruidFragment),
```

Comment:
Clone not required here
```suggestion
        DruidRole::Broker => druid
            .spec
            .brokers
            .role_groups
            .get(&rolegroup_ref.role_group)
            .map(|rg| &rg.config.config)
            .and_then(|rg| rg.resources.clone())
            .map(RoleResourceFragment::DruidFragment),
```